File: matlabpool_test.m
clear
close all

global N

N = 50; % count of jobs in a single test
MatlabPool.init(2);
MatlabPool.clear(); % cancel all jobs

count_eval.Small = 100;
count_eval.Normal = 30;
count_eval.Large = 5;
count_eval.Huge = 1;

i = 1;
while(true)
    name = sprintf('test_%d',i);
    fun = str2func(name);
    
    if(~exist(name,'file'))
        break
    end

    fprintf('run %9s...',name)
    fun();
    status = MatlabPool.statusJobs;
    if ~all(structfun(@(x)length(x),status) == 0)
        fprintf('there are jobs in the pool\n')
        error('there are jobs in the pool')
    end
    fprintf('passed\n')
    i = i+1;
end
fprintf('<strong> all test passed </strong>\n')

%% Test 1: sqrt(i) with i = 0,1,...,N, double
function test_1() %#ok
    global N
    for i = N:-1:1
        id(i) = MatlabPool.submit('sqrt',1,i);
    end
    for i = N:-1:1
        result = MatlabPool.wait(id(i));
        assert(abs(result{1}-sqrt(i)) < eps)
    end
end

%% Test 2: sqrt(i) with i = 0,1,...,N, float
function test_2() %#ok
    global N
    for i = N:-1:1
        id(i) = MatlabPool.submit('sqrt',1,single(i));
    end
    for i = N:-1:1
        result = MatlabPool.wait(id(i));
        assert(abs(result{1}-sqrt(single(i))) < eps)
    end
end

%% Test 3: increase/decrease pool size
function test_3() %#ok
    global N
    for i = N:-1:1
        id(i) = MatlabPool.submit('sqrt',1,i);
    end
    n = MatlabPool.size();
    MatlabPool.resize(n + 2);
    assert(n + 2 == MatlabPool.size())
    for i = N:-1:1
        result = MatlabPool.wait(id(i));
        assert(abs(result{1}-sqrt(i)) < eps)
    end
    
    for i = N:-1:1
        id(i) = MatlabPool.submit('sqrt',1,i);
    end
    MatlabPool.resize(n);
    assert(n == MatlabPool.size())
    for i = N:-1:1
        result = MatlabPool.wait(id(i));
        assert(abs(result{1}-sqrt(i)) < eps)
    end
end

%% Test 4: restart pool
function test_4() %#ok
    global N
    for i = N:-1:1
        MatlabPool.submit('sqrt',1,i);
    end
    
    n = MatlabPool.size();
    MatlabPool.shutdown();
    MatlabPool.init(n);
end

%% Test 5: clear pool
function test_5() %#ok
    global N
    for i = N:-1:1
        MatlabPool.submit('sqrt',1,i);
    end
    MatlabPool.clear();
end

%% Test 6: wait for undefined job
function test_6() %#ok
    try
        MatlabPool.wait(uint64(0));
    catch e
        if strcmp(e.identifier,'MatlabPoolMEX:JobNotExists')
            return
        end
    end
    error('no error thrown')
end

%% Test 7: wait x2 for same job
function test_7() %#ok
    id = MatlabPool.submit('sqrt',1,47);
    MatlabPool.wait(id);
    try
        MatlabPool.wait(id);
    catch e
        if strcmp(e.identifier,'MatlabPoolMEX:JobNotExists')
            return
        end
    end
    error('no error thrown')
end

%% Test 8: eval
function test_8() %#ok
    MatlabPool.eval('pwd');
end

%% Test 9: jobs and eval
function test_9() %#ok
    global N
    for i = N:-1:1
        id(i) = MatlabPool.submit('sqrt',1,i);
    end
    MatlabPool.eval('pwd');
    for i = N:-1:1
        result = MatlabPool.wait(id(i));
        assert(abs(result{1}-sqrt(i)) < eps)
    end
end

%% Test 10: invalid eval
function test_10() %#ok
    try
        MatlabPool.eval('pwwwd')
    catch e
        if strcmp(e.identifier,'MatlabPoolMEX:JobExecutionError')
            return
        end
    end
end

%% Test 11: invalid job
function test_11() %#ok
    id = MatlabPool.submit('sqqqrt',1,47);
    try
        MatlabPool.wait(id);
    catch e
        if strcmp(e.identifier,'MatlabPoolMEX:JobExecutionError')
            return
        end
    end
    error('no error thrown')
end

%% Test 12: job with disp
function test_12() %#ok
    id = MatlabPool.submit('disp',0,'Hello World');
    MatlabPool.wait(id);
end

%% Test 13: get job status
function test_13() %#ok
    global N
    for i = N:-1:1
        id(i) = MatlabPool.submit('sqrt',1,i);
    end
    status = MatlabPool.statusJobs;
    assert(length(intersect(fieldnames(status),{'Status','WorkerID','JobID'})) == 3)
    assert(all(structfun(@(x)length(x),status) == N))
    for i = N:-1:1
        result = MatlabPool.wait(id(i));
        assert(abs(result{1}-sqrt(i)) < eps)
    end
end

%% Test 14: cancel jobs
function test_14() %#ok
    global N
    time = 0.001;
    for i = N:-1:1
        id(i) = MatlabPool.submit('pause',1,time);
    end
    for i = 1:N
        MatlabPool.cancel(id(i));
    end
end

%% Test 15: get worker status
function test_15() %#ok
    global N
    for i = N:-1:1
        id(i) = MatlabPool.submit('sqrt',1,i);
    end
    status = MatlabPool.statusWorker;
    assert(length(intersect(fieldnames(status),{'Ready'})) == 1)
    assert(all(structfun(@(x)uint64(length(x)),status) == MatlabPool.size()))
    for i = N:-1:1
        result = MatlabPool.wait(id(i));
        assert(abs(result{1}-sqrt(i)) < eps)
    end
end

File: src/matlabpool_test.cpp
#include <iostream>
#include <exception>

#include "MatlabPool.hpp"
#include "TestSuite.hpp"

#include <queue>
#include <cmath>
#include <chrono>
#include <thread>

// TODO test mit valgrind
// TODO test C0 ueberdeckung

std::size_t run_tests()
{
    using Effort = TestSuite::Effort;
    using namespace MatlabPool;

    constexpr const std::size_t N = 50; // count of jobs in a single test

    TestSuite test;
    test.set_countEval(Effort::Small, 100);
    test.set_countEval(Effort::Normal, 30);
    test.set_countEval(Effort::Large, 5);
    test.set_countEval(Effort::Huge, 1);

    const unsigned int nof_worker = 2;
    std::vector<std::u16string> options = {u"-nojvm", u"-nosplash"};

    auto pool = std::unique_ptr<Pool>(PoolLibLoader::createPool(nof_worker, options));

    test.set_postFun([&]() {
        auto status = pool->get_job_status();
        std::size_t count_jobs = status[0]["JobID"].getNumberOfElements();
        Assert(0 == count_jobs, "there are jobs in the pool");
    });

    // Create  MATLAB data array factory
    matlab::data::ArrayFactory factory;

    test.run("sqrt(i) with i = 0,1,...,N, double", Effort::Normal, [&]() {
        using Float = double;
        std::array<JobID, N> jobid;
        std::vector<bool> worker_used(pool->size(), false);
        for (std::size_t i = 0; i < N; i++)
        {
            jobid[i] = pool->submit(
                JobFeval(u"sqrt", 1, {factory.createScalar<Float>(Float(i))}));
        }

        for (std::size_t i = 0; i < N; i++)
        {
            JobFeval job = pool->wait(jobid[i]);
            worker_used[job.get_workerID()] = true;
            Assert(1 == job.peek_result().size(), "unexpect size of result");
            matlab::data::TypedArray<Float> result = job.peek_result()[0];
            Assert(std::sqrt(Float(i)) == Float(result[0]), "unexpect result");
            Assert(job.get_outBuf().empty(), "output buffer should be empty");
            Assert(job.get_errBuf().empty(), "error buffer should be empty");
        }
        for (auto e : worker_used)
            Assert(e, "unused worker");
    });

    test.run("sqrt(i) with i = 0,1,...,N, float", Effort::Normal, [&]() {
        using Float = float;
        std::array<JobID, N> jobid;
        for (std::size_t i = 0; i < N; i++)
        {
            jobid[i] = pool->submit(
                JobFeval(u"sqrt", 1, {factory.createScalar<Float>(Float(i))}));
        }

        for (std::size_t i = 0; i < N; i++)
        {
            JobFeval job = pool->wait(jobid[i]);
            Assert(1 == job.peek_result().size(), "unexpect size of result");
            matlab::data::TypedArray<Float> result = job.peek_result()[0];
            Assert(std::sqrt(Float(i)) == Float(result[0]), "unexpect result");
        }
    });

    test.run("increase/decrease pool size", Effort::Huge, [&]() {
        std::array<JobID, N> jobid;
        // increase
        for (JobID &i : jobid)
        {
            i = pool->submit(
                JobFeval(u"pause", 0, {factory.createScalar<double>(0.01)}));
        }

        pool->resize(pool->size() + 2, options);
        for (JobID i : jobid)
            pool->wait(i);

        // decrease
        for (JobID &i : jobid)
        {
            i = pool->submit(
                JobFeval(u"pause", 0, {factory.createScalar<double>(0.01)}));
        }
        pool->resize(pool->size() - 2, options);
        for (JobID i : jobid)
            pool->wait(i);
    });

    test.run("restart pool", Effort::Huge, [&]() {
        std::array<JobID, N> jobid;
        for (JobID &i : jobid)
        {
            i = pool->submit(
                JobFeval(u"pause", 0, {factory.createScalar<double>(0.01)}));
        }

        pool = std::unique_ptr<Pool>(PoolLibLoader::createPool(nof_worker, options));
    });

    test.run("empty pool size", Effort::Small, [&]() {
        UnexpectException<Pool::EmptyPool>::check([&]() {
            pool->resize(0, options);
        });
    });

    test.run("wait for undefined job", Effort::Small, [&]() {
        UnexpectException<Pool::JobNotExists>::check([&]() {
            pool->wait(JobID(-1));
        });
    });

    test.run("wait x2 for same job", Effort::Small, [&]() {
        JobID id = pool->submit(
            JobFeval(u"sqrt", 1, {factory.createScalar<double>(0.5)}));

        pool->wait(id);
        UnexpectException<Pool::JobNotExists>::check([&]() {
            pool->wait(id);
        });
    });

    test.run("eval", Effort::Normal, [&]() {
        JobEval job(u"pwd");
        pool->eval(job);

        Assert(job.get_errBuf().empty(), "error buffer should be empty");
        Assert(!job.get_outBuf().empty(), "empty output buffer");
    });

    test.run("jobs and eval", Effort::Normal, [&]() {
        using Float = double;
        std::array<JobID, N> jobid;
        for (std::size_t i = 0; i < N; i++)
        {
            jobid[i] = pool->submit(
                JobFeval(u"sqrt", 1, {factory.createScalar<Float>(Float(i))}));
        }
        JobEval job_eval(u"pwd");
        pool->eval(job_eval);

        for (std::size_t i = 0; i < N; i++)
            pool->wait(jobid[i]);
    });

    test.run("invalid eval", Effort::Normal, [&]() {
        JobEval job(u"pwwd");

        pool->eval(job);

        Assert(job.get_outBuf().empty(), "output buffer should be empty");
        Assert(!job.get_errBuf().empty(), "empty error buffer");
    });

    test.run("invalid job", Effort::Normal, [&]() {
        JobID id = pool->submit(
            JobFeval(u"sqqqqrt", 1, {factory.createArray<double>({0})}));

        auto job = pool->wait(id);

        UnexpectException<JobBase::ExecutionError>::check([&]() {
            job.pop_result();
        });
        Assert(!job.get_errBuf().empty(), "empty error buffer");
    });

    test.run("job with disp", Effort::Normal, [&]() {
        JobID id = pool->submit(
            JobFeval(u"disp", 0, {factory.createCharArray("Hello World!")}));

        JobFeval job = pool->wait(id);

        Assert(job.get_errBuf().empty(), "error buffer should be empty");
        Assert(!job.get_outBuf().empty(), "empty output buffer");
    });

    test.run("get job status", Effort::Normal, [&]() {
        using Float = float;
        std::array<JobID, N> jobid;
        for (std::size_t i = 0; i < N; i++)
        {
            jobid[i] = pool->submit(
                JobFeval(u"sqrt", 1, {factory.createScalar<Float>(Float(i))}));
        }

        for (std::size_t i = 0; i < N; i++)
        {
            JobFeval job = pool->wait(jobid[i]);
            Assert(1 == job.peek_result().size(), "unexpect size of result");
            matlab::data::TypedArray<Float> result = job.peek_result()[0];
            Assert(std::sqrt(Float(i)) == Float(result[0]), "unexpect result");

            auto status = pool->get_job_status();

            Assert(N - i - 1 == status[0]["JobID"].getNumberOfElements(),
                   "unexpect size for \"JobID\" field");

            Assert(N - i - 1 == status[0]["Status"].getNumberOfElements(),
                   "unexpect size for \"Status\" field");

            Assert(N - i - 1 == status[0]["WorkerID"].getNumberOfElements(),
                   "unexpect size for \"WorkerID\" field");
        }
    });

    test.run("cancel all jobs", Effort::Normal, [&]() {
        using Float = float;
        std::array<JobID, N> jobid;
        for (std::size_t i = 0; i < N; i++)
        {
            jobid[i] = pool->submit(
                JobFeval(u"sqrt", 1, {factory.createScalar<Float>(Float(i))}));
        }

        pool->clear();
    });

    test.run("cancel jobs", Effort::Normal, [&]() {
        using Float = double;
        Float pause = 0.05;
        std::array<JobID, N> jobid;
        for (std::size_t i = 0; i < N; i++)
        {
            jobid[i] = pool->submit(
                JobFeval(u"pause", 0, {factory.createScalar<Float>(pause)}));
        }

        for (std::size_t i = 0; i < N; i++)
            pool->cancel(jobid[N - i - 1]);

        auto status = pool->get_job_status();
        matlab::data::TypedArray<JobID> jobid_field = status[0]["JobID"];

        if (jobid_field.getNumberOfElements() != 0)
            UnexpectCondition("unexpect jobs in pool");
    });

    test.run("cancel endless jobs", Effort::Normal, [&]() {
        using Float = double;
        Float pause = 1e10;
        std::array<JobID, N> jobid;
        for (std::size_t i = 0; i < N; i++)
        {
            jobid[i] = pool->submit(
                JobFeval(u"pause", 0, {factory.createScalar<Float>(pause)}));
        }

        for (std::size_t i = 0; i < N; i++)
            pool->cancel(jobid[N - i - 1]);

        auto status = pool->get_job_status();
        matlab::data::TypedArray<JobID> jobid_field = status[0]["JobID"];

        if (jobid_field.getNumberOfElements() != 0)
            UnexpectCondition("unexpect jobs in pool");
    });

    test.run("get worker status", Effort::Large, [&]() {
        using Float = double;
        JobID id = pool->submit(
            JobFeval(u"pause", 0, {factory.createScalar<Float>(Float(1))}));

        std::this_thread::sleep_for(std::chrono::milliseconds(500));

        matlab::data::TypedArray<bool> ready1 = pool->get_worker_status()[0]["Ready"];
        std::size_t count_busy = 0;
        for (bool b : ready1)
            if (!b)
                ++count_busy;

        Assert(count_busy == 1, "only on worker should be busy");
        pool->wait(id);
    });

    return test.get_failed();
}

int main()
{
    try
    {
        return int(run_tests());
    }
    catch (const std::exception &e)
    {
        std::cout << "Abort test: " << e.what() << std::endl;
        return 1;
    }
    catch (...)
    {
        std::cout << "Abort test" << std::endl;
        return 1;
    }
    return 0;
}